### PR TITLE
model: Add `mpackage` importation hierarchy

### PR DESCRIPTION
**Pull request by Delja**
Added a `mpackage_importation_hierarchy` graph to keep the hierarchy of packages. The packages relations are defined when the `set_imported_mmodules` is called. 

This pr is related to the programming by contract. The objective is to determine when the use of a contract is necessary in a given context.

Note in this graph all the packages are in relation with themself.

example for all nitc package hierarchy :

This pr has a dependency with the #2784

### CODE REVIEW

File: src/model/mmodule.nit [hunk 1]
```diff
@@ -186,11 +186,16 @@ class MModule
 	end
 
 	# Register the imported modules (ie "import some_module")
+	# In the same time it register the imported package
 	# The visibility must be set with `set_visibility_for`.
 	fun set_imported_mmodules(imported_mmodules: Array[MModule])
 	do
 		for m in imported_mmodules do
 			self.model.mmodule_importation_hierarchy.add_edge(self, m)
+			if self.mpackage isa MPackage and m.mpackage isa MPackage then
```
Comment (privat) [inline src/model/mmodule.nit]: how could mpackage not be a MPackage? Or do you mean `!= null` instead of `isa`?

File: src/model/mmodule.nit [hunk 2]
```diff
@@ -186,11 +186,16 @@ class MModule
...
+				# Register the imported package
+				self.model.mpackage_importation_hierarchy.add_arc(self.mpackage.as(MPackage), m.mpackage.as(MPackage))
```
Comment (privat) [inline src/model/mmodule.nit]: `isa` then `as` is a bad practice. use local variables and adaptive typing

File: src/model/mpackage.nit [hunk 3]
```diff
@@ -177,6 +178,11 @@ class MGroup
 end
 
 redef class Model
+
+	# Full package importation hierarchy
+	# Each package is in relation with itself
+	var mpackage_importation_hierarchy = new ReflexiveHashDigraph[MPackage]
```
Comment (privat) [inline src/model/mpackage.nit]: its more a graph than a hierarchy : no systematic transitivity and no anti-symmetry.

File: src/model/mmodule.nit (hunk 1 above)
Comment (Delja) [reply src/model/mmodule.nit]: Yes, it was to perform the null test. I changed it by using local variables and adaptive typing as you indicated.

File: src/model/mpackage.nit (hunk 3 above)
Comment (Delja) [reply src/model/mpackage.nit]: I see. I changed the name for `mpackage_importation_graph`

**Pull request merged**